Batteries/Data/Vector/Basic.lean

```diff
@@ -0,0 +1,507 @@
+
+
+
+
+
+
+/-- Make an empty vector with pre-allocated capacity-/
+def mkEmpty (capacity : Nat): Vector α 0 := ⟨Array.mkEmpty capacity, rfl⟩
+
+
+/-- The list obtained from a vector. -/
+def toList (v : Vector α n) : List α :=
+  v.1.1
+
+
+/-- nth element of a vector, indexed by a `Fin` type. -/
+def get (v : Vector α n) (i : Fin n) : α :=
+  v.1.get <| i.cast v.2.symm
+
+/-- `getD v i v₀` gets the `iᵗʰ` element of v if valid. Else it returns `v₀` by default-/
+abbrev getD (v : Vector α n) (i : Nat) (v₀ : α) : α :=
+  Array.getD v.toArray i v₀
+
+/--`get! v i` gets the `iᵗʰ` element of v if valid, else panics-/
+def get! [Inhabited α] (v : Vector α n) (i : Nat) : α :=
+  v.toArray.get! i
+
+/--`v.back!` gets the last element of the vector.
+It uses the panicing function `get!`-/
+def back! [Inhabited α] (v : Vector α n) : α :=
+  Vector.get! v (n - 1)
+
+/--`v.back?` gets the last element `x` of the array as `some x`
+if it exists. Else the vector is empty and it returns `none`-/
+def back? (v : Vector α n) : Option α :=
+  v.toArray.get? (n - 1)
+
+/-- `push v x` pushes `x` to the end of vector `v` in $$O(1)$$ time -/
+def push (x : α) (v : Vector α n)  : Vector α (n + 1) :=
+  ⟨v.toArray.push x, proof⟩
+  where
+    proof := by
+      simp only [size_push, Nat.reduceEqDiff]
+      exact v.size_eq
+      done
+
+/--Set an element in a vector without bounds checks, using a Fin index.
+
+This will perform the update destructively provided that a has a reference count of 1 when called.-/
+def set (v : Vector α n) (i : Fin n) (x : α) : Vector α n :=
+  ⟨v.toArray.set (v.size_eq.symm ▸ i) x, by {simp only [size_set]; exact v.size_eq}⟩
+
+/-- Set an element in a vector, or do nothing if the index is out of bounds.
+
+This will perform the update destructively provided that a has a reference count of 1 when called.-/
+def setD (v : Vector α n) (i : Nat) (x : α) : Vector α n :=
+  ⟨v.toArray.setD i x, proof⟩
+  where
+    proof := by
+      simp only [size_setD]
+      exact v.size_eq
+      done
+/--
+Set an element in an array, or panic if the index is out of bounds.
+
+This will perform the update destructively provided that a has a reference count of 1 when called.
+-/
+def set! (v : Vector α n) (i : Nat) (x : α) : Vector α n :=
+  ⟨v.toArray.set! i x, proof⟩
+  where
+    proof := by
+      simp; exact v.size_eq
+      done
+
+
+/-- Appending a vector to another. -/
+def append {n m : Nat} : Vector α n → Vector α m → Vector α (n + m)
+  | ⟨a₁, h₁⟩, ⟨a₂, h₂⟩ => ⟨a₁ ++ a₂, by rw[Array.size_append, ←h₁,←h₂]⟩
+
+
+instance : HAppend (Vector α n) (Vector α m) (Vector α (n + m)) where
+  hAppend := append
+/--
+  `Vector.head` produces the head of a vector
+-/
+def head (v : Vector α (n+1)) :=
+  v.toArray.get ⟨0, v.size_eq.symm ▸ Nat.zero_lt_succ n⟩
+
+
+/--
+  `Vector.cons` is the list cons pattern for vectors
+-/
+
+def cons (x : α) (xs : Vector α n) : Vector α (n + 1) := Nat.add_comm 1 n ▸ (#v[x] ++ xs)
```

Comment:
Backward compatibility with Mathlib's Vector is the only reason I can think of.  We could try turning it into an abbrev (ref : [the comment above](https://github.com/leanprover-community/batteries/pull/793#discussion_r1599040955))

Reply:
There's a danger here that a user might think this is an efficient op since `cons` is O(1) for lists. However, `push` is the `O(1)` op for arrays and vectors. There at least needs to be a big warning about this. (Especially since it is actually `O(1)` for mathlib vectors.)

Reply:
Here's a suggestion : We leave this definition in place for compatibility, but add a warning that this operation is not O(1). I think the docstrings support markdown so we can make the warning bold.

Reply:
OTOH, as I said in the zulip thread. This cons function is very unnatural for Vectors. No memory allocator expands an allocation bidirectionally in memory

Reply:
This is always a tough call. The first option is to add a big warning so that users hopefully don't use this function expecting it to be better than O(n). The second option is to prevent the footgun by not having this function at all. Since `#v[x] ++ v` is not less convenient than `.cons x v`, I lean toward option 2.

Reply:
I think for now this should be as it is. We will need a clean up PR to rearrange defs, add any missing lemmas and so on, after everything else about array has been moved.